File: problems/Misc/moleculesToAtoms.hs
-- For a given chemical formula represented by a string, count the number of atoms
-- of each element contained in the molecule and return a key value mapping of
-- elements to the number of them contained in the molecule

-- "H" -> Right [("H", 1)] -- Hydrogen
-- "O2" -> Right [("O", 2)] -- Oxygen
-- "H2O" -> Right [("H", 2), ("O", 1)] -- Water
-- "Mg(OH)2" -> Right [("Mg",1),("O",2),("H",2)] -- Magnesium hydroxide
-- "K4[ON(SO3)2]2" -> Right [("K",4),("O",14),("N",2),("S",4)] -- Fremy's salt
-- "pie" -> Left "Not a valid molecule"

import AParser
import Control.Applicative
import Data.Char

data Mol = A { name :: String, num :: Int }
         | B { mols :: [Mol], num :: Int }
  deriving (Show)

atomName :: Parser String
atomName = spaces *> ((:) <$> (satisfy isUpper) <*> (zeroOrOne $ satisfy isLower)) <* spaces

openBrac :: Parser Char
openBrac = spaces *> (char '(' <|> char '[') <* spaces

closeBrac :: Parser Char
closeBrac = spaces *> (char ')' <|> char ']') <* spaces

atom :: Parser Mol
atom = spaces *> (A <$> atomName <*> (posInt <|> pure 1))

block :: Parser Mol
block = B <$> insideBlock <*> posInt

insideBlock :: Parser [Mol]
insideBlock = openBrac *> (oneOrMore mol) <* closeBrac

mol :: Parser Mol
mol = atom <|> block

parseMols :: Parser [Mol]
parseMols = oneOrMore mol

combine :: [Mol] -> [(String, Int)]
combine = foldr (\ a as -> case a of
                  (A name n) -> (name, n):as
                  (B xs n)   -> f n (combine xs) ++ as) []
  where f n = map (\ (name, num) -> (name, num * n))

merge :: [(String, Int)] -> [(String, Int)]
merge = foldr (\ (s, n) xs -> case lookup s xs of
                                (Just on) -> (s, on + n) : [x | x <- xs, fst x /= s]
                                _         -> (s, n) : xs) []

convert (Just (parsed, _)) = Right $ merge $ combine parsed
convert _                  = Left "Not a valid molecule"

main = do
  formula <- getLine
  let parsed = runParser parseMols formula
  putStrLn $ show (convert parsed)

File: problems/Misc/moleculesToAtoms.py
def f(cs, es={}):
    if (len(cs) == 0):
        return es
    cs = atom(cs, es)
    (ncs, nes) = block(cs, es)
    return f(ncs, nes)

def atom(cs, es):
    (name, qty, cs) = parse_atom(cs)
    if(name):
        if(name in es):
            es[name] += qty
        else:
            es[name] = qty
    return cs

def parse_atom(cs):
    a = cs[0]
    b = cs[1]
    if(a.isupper()):
        if(b.islower()):
            return (a+b, int(cs[2]), cs[3:])
        elif(b.isnumeric()):
            return (a, int(b), cs[2:])
        return (a, 1, cs[1:])
    return (False, 2, cs)

def block(cs, es):
    if(len(cs) == 0):
        return (cs, es)
    a = cs[0]
    if(a == '('):
        (cs1, n, rest) = inside_block(cs[1:])
        return (rest, merge(es, scale(f(cs1, {}), n)))
    else:
        return (cs, es)

def merge(a, b):
    r = {}
    for k in a:
        if(k in b):
            r[k] = a[k] + b[k]
        else:
            r[k] = a[k]
    for k in b:
        if(k not in a):
            r[k] = b[k]
    return r

def inside_block(cs):
    c_index = mismatched(cs, 0, 0)
    return (cs[:c_index], int(cs[c_index + 1]), cs[c_index + 2:])

def scale(dic, n):
    r = {}
    for k in dic:
        r[k] = dic[k] * n
    return r

def mismatched(cs, count, i):
    a = cs[i]
    if(count < 0):
        return i - 1
    if(a == '('):
        return mismatched(cs, count+1, i+1)
    if(a == ')'):
        return mismatched(cs, count-1, i+1)
    else:
        return mismatched(cs, count, i+1)

print(f("H3(H2(O2)4)2"))
